### PR TITLE
Clarify comments describing how GuaranteedUpdate() (previously AtomicUpd...

### LINKED ISSUE #6626: Decide if current ETCD Update semantics is correct.
While having fun with subresources I found out that we have somewhat unexpected logic in etcd_helper.go AtomicUpdate (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/tools/etcd_helper.go#L311). Namely, IIUC, there's an inbuilt retry loop in case of precondition failure, so effectively we do not have an 'optimistic locking' on updated and each update will eventually go through.

I'm not saying that this is wrong - it just was a bit surprising for me, and other people in Warsaw as well. Is this by design, and everyone was aware of that? I'm asking before I'm afraid that it will cause race conditions and all the fun they generate.

@wojtek-t @thockin @lavalamp @davidopp @brendanburns

### PR DESCRIPTION
**Pull request by davidopp**
...ate() works. Closes #6626.

@wojtek-t

### CODE REVIEW

File: pkg/tools/etcd_helper.go [hunk 1]
```diff
@@ -284,31 +284,33 @@ func (h *EtcdHelper) SetObj(key string, obj, out runtime.Object, ttl uint64) err
...
+// deciding how the updated object (that it returns) should look.
 //
 // Example:
 //
 // h := &util.EtcdHelper{client, encoding, versioning}
-// err := h.AtomicUpdate("myKey", &MyType{}, true, func(input runtime.Object) (runtime.Object, uint64, error) {
-//	// Before this function is called, currentObj has been reset to etcd's current
-//	// contents for "myKey".
+// err := h.GuaranteedUpdate("myKey", &MyType{}, true, func(input runtime.Object) (runtime.Object, uint64, error) {
+//	// Before each invoation of the user-defined function, "input" is reset to etcd's current contents for "myKey".
```
Comment (wojtek-t) [inline pkg/tools/etcd_helper.go]: s/invoation/invocation/

File: pkg/tools/etcd_helper.go [hunk 2]
```diff
@@ -284,31 +284,33 @@ func (h *EtcdHelper) SetObj(key string, obj, out runtime.Object, ttl uint64) err
...
-// Pass an EtcdUpdateFunc to EtcdHelper.AtomicUpdate to make an atomic etcd update.
-// See the comment for AtomicUpdate for more detail.
+// Pass an EtcdUpdateFunc to EtcdHelper.GuaranteedUpdate to make an etcd update that is guaranteed to succeed.
+// See the comment for GuaranteedUpdate for more detail.
 type EtcdUpdateFunc func(input runtime.Object) (output runtime.Object, ttl uint64, err error)
 
-// AtomicUpdate generalizes the pattern that allows for making atomic updates to etcd objects.
-// Note, tryUpdate may be called more than once.
+// GuaranteedUpdate calls "tryUpdate()" to update key "key" that is of type "ptrToType". It keeps
+// retrying the update even in the face of etcd index conflict, until it succeeds. Note that object
```
Comment (wojtek-t) [inline pkg/tools/etcd_helper.go]: It retries only if  "IsEtcdTestFailed" returns true - IIUC this is the case only if there is precondition failure. So maybe it's worth mentioning that it retires only if there is precondition failure (or index conflict which I understood as exactly the same) and doesn't retry on any other errors?

**Comment by wojtek-t** [comment]
LGTM - just two nits.

File: pkg/tools/etcd_helper.go (hunk 1 above)
Comment (davidopp) [inline pkg/tools/etcd_helper.go]: Done.

File: pkg/tools/etcd_helper.go (hunk 2 above)
Comment (davidopp) [inline pkg/tools/etcd_helper.go]: Done.

**Comment by davidopp** [comment]
Pushed new commit, PTAL.

**Comment by wojtek-t** [comment]
LGTM - merging.

**Pull request merged**